src/slice/sort.cr

```diff
@@ -206,4 +206,357 @@ struct Slice(T)
+    if v.size >= 2 && cmp(v[1], v[0], comp) < 0
+      x, v[0] = v[0], v[1]
+      (2...v.size).each do |i|
+        if cmp(v[i], x, comp) < 0
+          v[i - 1] = v[i]
+        else
+          v[i - 1] = x
+          return
+        end
+      end
+      v[v.size - 1] = x
+    end
+  end
+
+  # Merges non-decreasing runs `v[..mid]` and `v[mid..]` using `buf` as temporary storage, and
+  # stores the result into `v[..]`.
+  protected def self.merge!(v, mid, buf, comp)
+    size = v.size
+
+    if mid <= size - mid
+      # The left run is shorter.
+      buf.copy_from(v.to_unsafe, mid)
+
+      left = 0
+      right = mid
+      out = v.to_unsafe
+
+      while left < mid && right < size
+        # Consume the lesser side.
+        # If equal, prefer the left run to maintain stability.
+        if cmp(v[right], buf[left], comp) < 0
+          out.value = v[right]
+          out += 1
+          right += 1
+        else
+          out.value = buf[left]
+          out += 1
+          left += 1
+        end
+      end
+
+      out.copy_from(buf + left, mid - left)
+    else
+      # The right run is shorter.
+      buf.copy_from((v + mid).to_unsafe, size - mid)
+
+      left = mid
+      right = size - mid
+      out = v.to_unsafe + size
+
+      while left > 0 && right > 0
+        # Consume the greater side.
+        # If equal, prefer the right run to maintain stability.
+        if cmp(buf[right - 1], v[left - 1], comp) < 0
+          left -= 1
+          out -= 1
+          out.value = v[left]
+        else
+          right -= 1
+          out -= 1
+          out.value = buf[right]
+        end
+      end
+
+      (v + left).copy_from(buf, right)
+    end
+  end
+
+  # Examines the stack of runs and identifies the next pair of runs to merge. More specifically,
+  # if `r` is returned, that means `runs[r]` and `runs[r + 1]` must be merged next. If the
+  # algorithm should continue building a new run instead, `nil` is returned.
+  #
+  # TimSort is infamous for its buggy implementations, as described here:
+  # http://envisage-project.eu/timsort-specification-and-verification/
+  #
+  # The gist of the story is: we must enforce the invariants on the top four runs on the stack.
+  # Enforcing them on just top three is not sufficient to ensure that the invariants will still
+  # hold for *all* runs in the stack.
+  #
+  # This function correctly checks invariants for the top four runs. Additionally, if the top
+  # run starts at index 0, it will always demand a merge operation until the stack is fully
+  # collapsed, in order to complete the sort.
+  @[AlwaysInline]
+  protected def self.collapse(runs)
+    n = runs.size
+    if n >= 2 &&
+       (runs[n - 1].begin == 0 ||
+       runs[n - 2].size <= runs[n - 1].size ||
+       (n >= 3 && runs[n - 3].size <= runs[n - 2].size + runs[n - 1].size) ||
+       (n >= 4 && runs[n - 4].size <= runs[n - 3].size + runs[n - 2].size))
+      n >= 3 && runs[n - 3].size < runs[n - 1].size ? n - 3 : n - 2
```

Comment:
this line so hard to read, mix of `if` and ternary operator

Reply:
I think so too. However, these lines are also ported from Rust completely (See [here](https://github.com/rust-lang/rust/blob/507bff92fadf1f25a830da5065a5a87113345163/library/alloc/src/slice.rs#L1132-L1137)). I'm worried whether this should be fixed or not in terms of maintainability.

@asterite or anyone, which do you like?

Reply:
In my experience leaving the code as similar as possible to the original one is better. Later on if the Rust code changes you can do a diff and see what changes you need to apply here. So please leave it like this.

And thank you for all your contributions, you are awesome! And I hope you are having fun coding in Crystal.